File: ios/App/SDK/BackendConfig.swift
// ---------------------------------------------------------------------------
// BackendConfig — user-entered backend descriptor + pure resolution to the SDK's
// two transport inputs. Native-owned (no KMP push). The /api/v1/ws path is the
// gateway contract; the security→allowSelfSigned mapping is a security boundary.
// Mirrors the Android BackendConfig (which is unit-tested); kept in sync.
// ---------------------------------------------------------------------------
import Foundation

/// Gateway WS path — the wire contract with the gateway. Not user-editable.
private let gatewayWSPath = "/api/v1/ws"

enum ConnectionSecurity: String, CaseIterable, Sendable {
    case tlsValid, tlsTrustSelfSigned, plainWs
}

struct BackendConfig: Equatable, Sendable {
    let host: String
    let port: Int
    let security: ConnectionSecurity

    var gatewayWsURL: String {
        let scheme = security == .plainWs ? "ws" : "wss"
        return "\(scheme)://\(host):\(port)\(gatewayWSPath)"
    }
    var allowSelfSigned: Bool { security == .tlsTrustSelfSigned }
}

enum ResolvedBackend: Equatable, Sendable {
    case configured(gatewayWsURL: String, allowSelfSignedDevHost: Bool)
    case unconfigured
}

/// Precedence: override → non-empty build-time default → unconfigured.
func resolveBackend(override: BackendConfig?, buildTimeDefaultURL: String, buildTimeAllowSelfSigned: Bool) -> ResolvedBackend {
    if let o = override { return .configured(gatewayWsURL: o.gatewayWsURL, allowSelfSignedDevHost: o.allowSelfSigned) }
    if !buildTimeDefaultURL.isEmpty { return .configured(gatewayWsURL: buildTimeDefaultURL, allowSelfSignedDevHost: buildTimeAllowSelfSigned) }
    return .unconfigured
}

File: android/src/main/kotlin/io/sentient/android/backend/BackendConfig.kt
// ---------------------------------------------------------------------------
// BackendConfig — the user-entered backend descriptor + the pure resolution
// logic that turns it (or the build-time default) into the SDK's two transport
// inputs. Native-owned per the design (no KMP push). Pure + unit-tested:
// the /api/v1/ws path is the gateway contract, and the security→allowSelfSigned
// mapping is a security boundary.
// ---------------------------------------------------------------------------
package io.sentient.android.backend

/** Gateway WS path — the wire contract with the gateway. Not user-editable. */
private const val GATEWAY_WS_PATH = "/api/v1/ws"

/** How the client secures the connection. Maps to scheme + self-signed trust. */
enum class ConnectionSecurity { TLS_VALID, TLS_TRUST_SELF_SIGNED, PLAIN_WS }

/**
 * A user-entered backend. [port] is validated by the caller (1..65535); [host]
 * is a bare host or IP (no scheme, no path).
 */
data class BackendConfig(
    val host: String,
    val port: Int,
    val security: ConnectionSecurity,
) {
    fun toGatewayWsUrl(): String {
        val scheme = if (security == ConnectionSecurity.PLAIN_WS) "ws" else "wss"
        return "$scheme://$host:$port$GATEWAY_WS_PATH"
    }

    /** True only for the explicit trust-self-signed selection. */
    fun allowSelfSigned(): Boolean = security == ConnectionSecurity.TLS_TRUST_SELF_SIGNED
}

/** The resolved transport inputs, or the signal to force the setup page. */
sealed interface ResolvedBackend {
    data class Configured(
        val gatewayWsUrl: String,
        val allowSelfSignedDevHost: Boolean,
    ) : ResolvedBackend
    data object Unconfigured : ResolvedBackend
}

/**
 * Precedence: runtime [override] → non-empty [buildTimeDefaultUrl] → Unconfigured.
 * [buildTimeAllowSelfSigned] applies only to the build-time-default branch
 * (debug trusts the local dev cert); an override carries its own trust posture.
 */
fun resolveBackend(
    override: BackendConfig?,
    buildTimeDefaultUrl: String,
    buildTimeAllowSelfSigned: Boolean,
): ResolvedBackend {
    if (override != null) {
        return ResolvedBackend.Configured(override.toGatewayWsUrl(), override.allowSelfSigned())
    }
    if (buildTimeDefaultUrl.isNotEmpty()) {
        return ResolvedBackend.Configured(buildTimeDefaultUrl, buildTimeAllowSelfSigned)
    }
    return ResolvedBackend.Unconfigured
}
